src/hii_water.py

```diff
@@ -35,139 +40,133 @@ class HIIWater(HIITask):
+            ),
+            "river_width_expand": ee.Kernel.euclidean(
+                radius=self.wide_river_width * 2, units="meters"
+            ),
+        }
+
+    def gpw_earliest(self):
+        return self.gpw.sort("system:time_start").first()
+
+    def gpw_latest(self):
+        return self.gpw.sort("system:time_start", False).first()
+
+    def gpw_interpolated(self):
+        gpw_prev = self.gpw.filterDate(
+            self.ee_taskdate.advance(-self.gpw_cadence, "year"), self.ee_taskdate
+        ).first()
+        gpw_next = self.gpw.filterDate(
+            self.ee_taskdate, self.ee_taskdate.advance(self.gpw_cadence, "year")
+        ).first()
+
+        gpw_delta_days = gpw_next.date().difference(gpw_prev.date(), "day")
+        taskdate_delta_days = self.ee_taskdate.difference(gpw_prev.date(), "day")
+
+        gpw_diff = gpw_next.subtract(gpw_prev)
+
+        gpw_daily_change = gpw_diff.divide(gpw_delta_days)
+        gpw_change = gpw_daily_change.multiply(taskdate_delta_days)
+
+        return gpw_prev.add(gpw_change)
 
     def calc(self):
-        scale = 300
-        gpw_cadence = 5
-        wide_river_width = 30  # meters
-        coast_settle_min = 4000  # meters
-        coast_settle_max = 15000  # meters
-        DECAY_CONSTANT = -0.0002
-        INDIRECT_INFLUENCE = 4
-        MIN_PIX_COUNT = 50
-        THRESHOLD_SCALE = 900
-
-        gpw, gpw_date = self.get_most_recent_image(
-            ee.ImageCollection(self.inputs["gpw"]["ee_path"])
-        )
-        watermask = ee.Image(self.inputs["watermask"]["ee_path"])
-        caspian = (
-            ee.Image(0)
-            .clip(ee.FeatureCollection(self.inputs["caspian"]["ee_path"]))
-            .unmask(1)
-        )
-        ocean = (
-            ee.Image(self.inputs["ocean"]["ee_path"])
-            .eq(0)
-            .add(caspian.eq(0).reproject(crs=self.crs, scale=self.scale))
-        )
-        jrc = (
-            ee.Image(self.inputs["jrc"]["ee_path"])
-            .select("occurrence")
-            .lte(70)
-            .unmask(1)
-            .multiply(caspian)
+        gpw_taskdate = None
+        ee_taskdate_millis = self.ee_taskdate.millis()
+        gpw_first_date = ee.Date(
+            self.gpw.sort("system:time_start").first().get("system:time_start")
+        ).millis()
+        gpw_last_date = ee.Date(
+            self.gpw.sort("system:time_start", False).first().get("system:time_start")
+        ).millis()
+        start_test = ee_taskdate_millis.lt(gpw_first_date)
+        end_test = ee_taskdate_millis.gt(gpw_last_date)
+        interpolate_test = start_test.eq(0).And(end_test.eq(0))
+        if interpolate_test.getInfo():
+            gpw_taskdate = self.gpw_interpolated()
+        elif end_test.getInfo():
+            gpw_taskdate = self.gpw_latest()
+        elif start_test.getInfo():
+            gpw_taskdate = self.gpw_earliest()
+        else:
+            raise Exception("no valid GPW image")
```

Comment:
I haven't looked, but I assume this is all the same as the GPW code we have elsewhere, right? If so we should make a card to refactor it out into `task_base`

Reply:
I added a card to trello